base/src/main/scala/data/ProCompose.scala

```diff
@@ -0,0 +1,54 @@
+package scalaz
+package data 
+
+object ProCompose {
+  import typeclass.{Profunctor, Strong, Category, StrongClass, CategoryClass}
+  import typeclass.Strong._
+  import typeclass.Profunctor._
+
+  trait ProCompose[P[_, _], A, B] { } 
+
+  case class Pure[P[_, _], A, B](pab: A => B) extends ProCompose[P, A, B]
+
+  case class Composed[P[_, _], A, X, B](pab: P[A, X], more: ProCompose[P, X, B]) extends ProCompose[P, A, B]
```

Comment:
This is `Free2` over `Prearrow` (`Semigroupoid` (compose) + `arr` (injection from function), see Tony's proposed type class hierarchy). As well, the lifting injection `P ~~> ProCompose[P, ?, ?]` is unnecessarily complicated and inefficient: `(pab: P[A, B]) => Composed(pab, Pure(identity[B]))`. I recommend creating an `Ops` trait with cases for `P[A, B]` and `A => B`, and making `Composed` reference it instead of `P[_, _]`; if variants of this structure will be common later, you may get more mileage out of implementing this as `TypeAlignedList[Ops]`.

More importantly, there is no stack-safe usage of this structure; unless you introduce an `Evaluable`-like type class:
```scala
  trait Evaluable[F[_, _], G[_]] {
    def eval[A, B](a: G[A], arr: F[A, B]): G[B]
  }
```